sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```diff
@@ -73,132 +79,167 @@
-          options.bucket, options.filterJson, options.parser, coder));
+    public <X> Read<X> withParser(Parser<X> f, Coder<X> coder) {
+      Preconditions.checkNotNull(f, "Parser cannot be null");
+      //coder can be null as it can be set on the output directly
+      return new Read<X>(new BoundedGridFSSource(options.uri, options.database,
+          options.bucket, options.filterJson, null), f, coder, allowedTimestampSkew);
+    }
+    public Read<T> allowedTimestampSkew(Duration skew) {
+      return new Read<T>(new BoundedGridFSSource(options.uri, options.database,
+          options.bucket, options.filterJson, null),
+          parser, coder, skew == null ? Duration.ZERO : skew);
     }
 
     public Read<T> withQueryFilter(String filterJson) {
-      return new Read<T>(new BoundedGridFSSource<T>(options.uri, options.database,
-          options.bucket, filterJson, options.parser, options.coder));
+      return new Read<T>(new BoundedGridFSSource(options.uri, options.database,
+          options.bucket, filterJson, null), parser, coder, allowedTimestampSkew);
     }
 
-    private final BoundedGridFSSource<T> options;
+    private final BoundedGridFSSource options;
+    private final Parser<T> parser;
+    private final Coder<T> coder;
+    private final Duration allowedTimestampSkew;
 
-    Read(BoundedGridFSSource<T> options) {
+    Read(BoundedGridFSSource options, Parser<T> parser,
+        Coder<T> coder, Duration allowedTimestampSkew) {
       this.options = options;
+      this.parser = parser;
+      this.allowedTimestampSkew = allowedTimestampSkew;
+      this.coder = coder;
+    }
+
+    BoundedGridFSSource getSource() {
+      return options;
     }
 
     @Override
     public PCollection<T> apply(PBegin input) {
-      org.apache.beam.sdk.io.Read.Bounded<T> unbounded =
+      org.apache.beam.sdk.io.Read.Bounded<ObjectId> bounded =
           org.apache.beam.sdk.io.Read.from(options);
-      PCollection<T> output = input.getPipeline().apply(unbounded);
-      if (options.coder != null) {
-        output.setCoder(options.coder);
+      PCollection<T> output = input.getPipeline().apply(bounded)
+          .apply(ParDo.of(new DoFn<ObjectId, T>() {
+            Mongo mongo;
+            GridFS gridfs;
+
+            @Setup
+            public void setup() {
+              mongo = options.setupMongo();
+              gridfs = options.setupGridFS(mongo);
+            }
+
+            @Teardown
+            public void teardown() {
+              mongo.close();
+            }
+
+            @ProcessElement
+            public void processElement(final ProcessContext c) throws IOException {
+              ObjectId oid = c.element();
+              GridFSDBFile file = gridfs.find(oid);
+              parser.parse(file, new ParserCallback<T>() {
+                @Override
+                public void output(T output, Instant timestamp) {
+                  if (timestamp == null) {
```

Comment:
Why not just checkNotNull(timestamp)? If the user didn't want to provide a timestamp, they'd call the other function instead.